core/src/smartcontracts/isi/domain.rs

```diff
@@ -47,28 +47,48 @@ pub mod isi {
             authority: &AccountId,
             state_transaction: &mut StateTransaction<'_, '_>,
         ) -> Result<(), Error> {
-            let account: Account = self.object.build(authority);
-            let account_id = account.id().clone();
+            let account_id = self.object.id().clone();
 
             if *account_id.domain_id() == *iroha_genesis::GENESIS_DOMAIN_ID {
                 return Err(InstructionExecutionError::InvariantViolation(
                     "Not allowed to register account in genesis domain".to_owned(),
                 ));
             }
 
-            let domain = state_transaction.world.domain_mut(&account_id.domain_id)?;
-            if domain.accounts.contains_key(&account_id) {
-                return Err(RepetitionError {
-                    instruction_type: InstructionType::Register,
-                    id: IdBox::AccountId(account_id),
+            match state_transaction.world.account_mut(&account_id) {
+                Ok(existing_account) => {
+                    if existing_account.is_active {
+                        return Err(RepetitionError {
+                            instruction_type: InstructionType::Register,
+                            id: IdBox::AccountId(account_id),
+                        }
+                        .into());
+                    }
+
+                    existing_account.activate();
```

Comment:
this part is now weird because `NewAccount` can carry more information (like some metadata) that you just disregard